### Expressions and Values: Arithmetic in Python<br>
* We're familiar with mathematical expressions like $3+4$ and $2-3/5$. Each expression is built out of
    * Values like $2$, $3$, and $5$ 
    * $\rm\color{orange}{Operators}$ (運算子) like $+$ and $-$, which combine their $\rm\color{orange}{operands}$ (運算元) in different ways  
    * In the expression $4/5$, the operator is $/$ and the operands are $4$ and $5$  
* Expressions do not have to involve an operator: A number by itself is an expression
    * For example, we consider $212$ to be an expression as well as a value  
    * Like any programming language, Python can evaluate basic mathematical expressions

In [ ]:
print(4 + 13)

* Here we typed print($4+13$), and then we pressed the Run key in order to signal that we were done entering that $\rm\color{orange}{expression}$
* Python then evaluated the expression
* When an expression is evaluated, it produces a single value. In the previous expression, the evaluation of $4+13$ produced the value $17$
* We use $\rm\color{orange}{print()}$ to show the value that is produced on the screen

In [ ]:
print(15 - 3)

In [ ]:
print(4 * 7)

In [ ]:
print(5 / 2)

* The result has a decimal point. In fact, the result of division $\rm\color{orange}{always}$ has a decimal point even if the result is a whole number

In [ ]:
print(4 / 2)

### Types <br>
* Every value in Python has a particular type, and the types of values determine how they behave when they are combined
* Values like $4$ and $17$ have type $\rm\color{orange}{int}$ (short for integer), and values like $2.5$ and $17.0$ have type $\rm\color{orange}{float}$
* The word float is short for floating point, which refers to the decimal point that moves around
between digits of the number
* An expression involving two floats produces a float

In [ ]:
print(17.0 - 10.0)

* When an expression’s operands are an int and a float, Python automatically converts the int to a float
* This is why the following two expressions both return the same answer

In [ ]:
print(17.0 - 10)
print(17 - 10.0)

* If you want, you can omit the zero after the decimal point when writing a floating-point number
* However, most people think this is bad style, since it makes your programs harder to read

In [ ]:
print(17. - 10)
print(17 - 10.)

### Integer Division, Modulo, and Exponentiation<br>
* Every now and then, we want only the integer part of a division result
* For example, we might want to know how many $24$-hour days there are in $53$ hours
* To calculate the number of days, we can use $\rm\color{orange}{integer\space division}$

In [6]:
print(53 // 24)

2


* We can find out how many hours are left over using the $\rm\color{orange}{modulo}$ operator, which gives the remainder of the division

In [7]:
print(53 % 24)

5


* Be careful about using $\%$ and $//$ with negative operands
* Python takes the floor of the result of an integer division, the result is one smaller than you might expect if the result is negative

In [8]:
print(-17 // 10)

-2


* When using modulo, the sign of the result matches the sign of the divisor (除數)

In [9]:
print(-17 % 10)
print(17 % -10)

3
-3


* Floating-point numbers can be operands for $//$ and $\%$ as well
* With $//$, division is performed and the result is rounded down to the nearest whole number, although the type is a floating-point number

In [10]:
print(3.3 // 1)
print(3 // 1.0)
print(3 // 1.1)
print(3.5 // 1.1)
print(3.5 // 1.3)

3.0
3.0
2.0
3.0
2.0


* The following expression calculates $3$ raised to the $6$th power

In [11]:
print(3 ** 6)

729


### What Is a Type?<br>
* We have seen two types of numbers (integers and floating-point numbers), so we ought to explain what we mean by a type
* In Python, a type consists of two things:
    * A set of values
    * A set of operations that can be applied to those values
---  
| Symbol | Operator | Example | Result |
| :--: | :--: | :--: | :--: |
| $-$ | Negation | $-5$ | $-5$ |
| $+$ | Addition | $11+3.1$ | $14.1$ |
| $-$ | Subtraction | $5-19$ | $-14$ |
| $*$ | Multiplication | 8.5 * 4 | $34.0$ |
| $/$ | Division | 11 / 2 | $5.5$ |
| $//$ | Integer Division | 11 // 2 | $5$ |
| $\%$ | Remainder | 8.5 % 3.5 | $1.5$ |
| $**$ | Exponentiation | 2 ** 5 | $32$ |

### Finite Precision<br>
* Floating-point numbers are not exactly the fractions you learned in grade school

In [12]:
print(2 / 3)
print(5 / 3)

0.6666666666666666
1.6666666666666667


* The first value ends with a $6$, and the second with a $7$
* This is fishy: Both of them should have an infinite number of $6$s after the decimal point
* The problem is that computers have a finite amount of memory
    * To make calculations fast and memory efficient, most programming languages limit how much information can be stored for any single number
    * The number $0.6666666666666666$ turns out to be the closest value to $2⁄3$ that the computer can actually store in that limited amount of memory
    * $1.6666666666666667$ is as close as we get to the real value of $5⁄3$

### More on Numeric Precision<br>
* Integers (values of type $\rm\color{orange}{int}$) in Python can be as large or as small as you like
* Float values are only $\rm\color{orange}{approximations}$ to real numbers
* Using more memory will $\rm\color{orange}{NOT}$ solve the problem
    * It will make the approximation closer to the real value
    * Writing a larger number of $6$s after the 0 in $0.666$… does not make it exactly equal to $2⁄3$
* The difference between $2/3$ and $0.6666666666666666$ may look tiny, but if we use $0.6666666666666666$ in a calculation, then the error may get compounded

In [13]:
print(2 / 3 + 1)
print(5 / 3)

1.6666666666666665
1.6666666666666667


* As we do more calculations, the rounding errors can get larger and larger
* Particularly if we are mixing very large and very small numbers

In [14]:
print(10000000000 + 0.00000000001)

10000000000.0


* The result ought to have twenty zeros between the first and last significant digit
    * That is too many for the computer to store
    * The result is just $10000000000$, as if the addition never took place
* Adding lots of small numbers to a large one can therefore have no effect at all
    * $\rm\color{orange}{Not}$ what a bank wants when it totals up the values of its customers’ savings accounts
* It is important to be aware of the floating-point issue
* There is no magic bullet to solve it, because computers are limited in both memory and speed
---
* Numerical analysis, the study of algorithms to approximate continuous mathematics, is one of the largest
subfields of computer science and mathematics
* $\rm\color{orange}{Tip}$: If you have to add up floating-point numbers, add them from smallest to largest in order to minimize the error

### Operator Precedence<br>
* Let’s put our knowledge of ints and floats to use in converting Fahrenheit to Celsius

In [ ]:
print(212 - 32 * 5 / 9)

* Python claims the result is $194.22222222222223$ degrees Celsius, when in fact it should be $100$
* The problem is that multiplication and division have higher precedence than subtraction
* When an expression contains a mix of operators, the $*$ and $/$ are evaluated before the $-$ and $+$
* Operators with higher precedence are applied before those with lower precedence
---
| Precedence | Operator | Operation |
| :--: | :--: | :--: |
| Highest | $**$ | Exponentiation |
|  | $-$ | Negation |
|  | $*$, $/$, $//$, $\%$ | Multiplication, division, integer division, and remainder |
| Lowest | $+$, $-$ | Addition and subtraction |

---
* Operators on the same row have equal precedence and are applied left to right, except for exponentiation, which is applied right to left

### Variables and Computer Memory: Remembering Values<br>
* Like mathematicians, programmers frequently name values so that they can use them later
* A name that refers to a value is called a $\rm\color{orange}{variable}$
* In Python, variable names can use letters, digits, and the underscore symbol (but they cannot start with a digit)
    * X, species $5618$, and degrees_celsius are all allowed
    * $777$ and no-way $!$ are not
* We create a new variable by $\rm\color{orange}{assigning}$ it a value

In [ ]:
degrees_celsius = 26.0

* This statement is called an $\rm\color{orange}{assignment}$ (賦值) statement
* We say that degrees_celsius is assigned the value $26.0$
* That makes degrees_celsius refer to the value $26.0$
* We can use variables anywhere we can use values
* Whenever Python sees a variable in an expression, it substitutes the value to which the variable refers

In [15]:
degrees_celsius = 26.0
print(degrees_celsius)
print(9 / 5 * degrees_celsius + 32)
print(degrees_celsius / degrees_celsius)

26.0
78.80000000000001
1.0


* Variables are called variables because their values can vary as the program executes
* We can assign a new value to a variable

In [ ]:
degrees_celsius = 26.0
print(9 / 5 * degrees_celsius + 32)
degrees_celsius = 0.0
print(9 / 5 * degrees_celsius + 32)

* Assigning a value to a variable that already exists does not create a second variable
* Instead, the existing variable is reused, which means that the variable no longer refers to its old value

In [ ]:
degrees_celsius = 15.5
difference = 100 - degrees_celsius
print(difference)

### Warning: $=$ Is Not Equality in Python!<br>
* In mathematics, $=$ means "the thing on the left is equal to the thing on the right"
* In Python, it means something quite different: $\rm\color{orange}{Assignment}$
* Assignment is not symmetric: $x=12$ assigns the value $12$ to variable $x$, but $12=x$ results in an error
* Because of this, we never describe the statement $x=12$ as "$x$ equals $12$." Instead, we read this as "$x$ gets $12$" or "$x$ is assigned $12$"

### Values, Variables, and Computer Memory<br>
* We are going to develop a memory model that will let us trace what happens when Python executes a Python program
* The model will help us accurately predict and explain what Python does when it executes code
* This skill is a requirement for becoming a good programmer

* Every location in the computer’s memory has a memory address that uniquely identifies that location
    * Much like an address for a house on a street
* We are going to mark our memory addresses with an $\rm\color{orange}{id}$ prefix (shor for identifier)
    * They look different from integers: *id1*, *id2*, *id3*, and so on
    * A memory address may look like: 0x001207
* Here is how we draw the floating-point value $26.0$ using the memory model
  
![Assignment01](lec01-01.jpg)  
  
* This picture shows the value $26.0$ at the memory address *id1*
    * We will always show the type of the value as well. In this case, float
* We will call this box an $\rm\color{orange}{object}$: A value at a memory address with a type
* During execution of a program, every value that Python keeps track of is stored inside an object in computer memory

* A variable contains the memory address of the object to which it refers
  
![Assignment02](lec01-02.jpg)  
  
* We usually draw arrows from variables to their objects to make the picture easier to interpret
* We use the following terminology
    * Value $26.0$ has the memory address *id1*
    * The object at the memory address *id1* has type float and the value $26.0$
    * Variable degrees_celsius $\rm\color{orange}{contains}$ the memory address *id1*
    * Variable degrees_celsius $\rm\color{orange}{refers}$ to the value $26.0$
---
* Whenever Python needs to know which value degree_celsius refers to, it looks at the object at the memory address that degree_celsius contains
* In this example, that memory address is *id1*, so Python will use the value at the memory address *id1*, which is $26.0$

### Assignment Statement<br>
* The general form of an assignment statement is
  
        «variable» = «expression» 
  
* It is executed as
    1. Evaluate the expression on the right of the $=$ sign to produce a value. This value has a memory address
    2. Store the memory address of the value in the variable on the left of the $=$
    3. Create a new variable if that name does not already exist; otherwise, just reuse the existing variable, replacing the memory address that it contains

In [16]:
degrees_celsius = 26.0 + 5
print(degrees_celsius)

31.0


* Here is how Python executes the statement degrees_celsius$=26.0+5$:
    1. Evaluate the expression on the right of the $=$ sign: $26.0+5$. This produces the value 31.0, which has a memory address (Remember that Python stores all values in computer memory)
    2. Make the variable on the left of the $=$ sign, degrees_celsius, refer to $31.0$ by storing the memory address of $31.0$ in degrees_celsius

### Reassigning to Variables<br>

In [17]:
difference = 20
double = 2 * difference
print(double)
difference = 5
print(double)

40
40


* This code demonstrates that assigning to a variable $\rm\color{orange}{does\space not\space change\space any\space other\space variable}$
    * We start by assigning value $20$ to variable *difference*
    * And then we assign the result of evaluating $2*$ *difference* (which produces $40$) to variable *double*
    * Next, we assign value $5$ to variable *difference*, but when we examine the value of *double*, it still refers to $40$

* Here is how it works according to our rules: The first statement, difference = $20$, is executed as follows:
    1. Evaluate the expression on the right of the $=$ sign: $20$. 
        * This produces the value $20$, which we will put at memory address *id1*
    2. Make the variable on the left of the $=$ sign, *difference*, refer to $20$ by storing *id1* in *difference*  
  
![Assignment03](lec01-03.jpg)  
  
* Variable double has not yet been created because we have not yet executed the assignment to it

* The second statement, *double*$=2*$ *difference*, is executed as follows:
    1. Evaluate the expression on the right of the $=$ sign: $2*$ *difference*
        * As we see in the memory model, *difference* refers to the value $20$
        * So this expression is equivalent to $2*20$, which produces $40$
        * We will pick the memory address *id2* for the value $40$
    2. Make the variable on the left of the $=$ sign, *double*, refer to $40$ by storing *id2* in *double*  
  
![Assignment04](lec01-04.jpg)

* When Python executes the third statement, it merely looks up the value that double refers to ($40$) and displays it
---
* The fourth statement, *difference* $=5$, is executed as follows:
    1. Evaluate the expression on the right of the $=$ sign: $5$
        * This produces the value $5$, which we’ll put at the memory address *id3*
    2. Make the variable on the left of the $=$ sign, difference, refer to $5$ by storing *id3* in *difference*  
  
![Assignment05](lec01-05.jpg)  
  
* Variable double still contains *id2*, so it still refers to $40$. Neither variable refers to $20$ anymore
---
* The fifth and last statement merely looks up the value that *double* refers to, which is still $40$, and displays it

In [18]:
number = 3
print(number)
number = 2 * number
print(number)
number = number * number
print(number)

3
6
36


### Augmented Assignment<br>
* In the following example, the variable score appears on both sides of the assignment statement

In [19]:
score = 50
print(score)
score = score + 20
print(score)

50
70


* This is so common that Python provides a shorthand notation for this operation

In [ ]:
score = 50
print(score)
score += 20
print(score)

* An augmented assignment combines an assignment statement with an operator to make the statement more concise
* An augmented assignment statement is executed as follows:
    1. Evaluate the expression on the right of the $=$ sign to produce a value
    2. Apply the operator attached to the $=$ sign to the variable on the left of the $=$ and the value that was produced 
    3. This produces another value. Store the memory address of that value in the variable on the left of the $=$
* Note that the operator is applied $\rm\color{orange}{after}$ the expression on the right is evaluated

In [24]:
d = 2
d *= 3 + 4
print(d)

14


In [ ]:
number = 10
number *= number
print(number)

# This code is equivalent to
number = 10
number = number * number
print(number)

| Symbol | Example | Result |
| :--: | :--: | :--: |
| $+=$ | x=7<br>x+=2 | $x$ refers to $9$ |
| $-=$ | x=7<br>x-=2 | $x$ refers to $5$ |
| $*=$ | x=7<br>x*=2 | $x$ refers to $14$ |
| $/=$ | x=7<br>x/=2 | $x$ refers to $3.5$ |
| $//=$ | x=7<br>x//=2 | $x$ refers to $3$ |
| $\%=$ | x=7<br>x\%=2 | $x$ refers to $1$ |
| $**=$ | x=7<br>x**=2 | $x$ refers to $49$ |

### How Python Tells You Something Went Wrong<br>
* Broadly speaking, there are two kinds of errors in Python
    1. $\rm\color{orange}{Syntax}$ (語法) errors: Happen when you type something that is not valid Python code
    2. $\rm\color{orange}{Semantic}$ (語意) errors: Happen when you tell Python to do something that it just cannot do
        * For example, divide a number by zero or try to use a variable that does not exist

In [ ]:
print(214 + Valentine)

* This is pretty cryptic; Python error messages are meant for people who already know Python
* The first two lines are not much use right now, though they will be indispensable when we start writing longer programs
* The last line is the one that tells us what went wrong: The name *Valentine* was not recognized

In [ ]:
print(2 + )

* The rules governing what is and is not legal in a programming language are called its $\rm\color{orange}{syntax}$
* The message tells us that we violated Python’s syntax rules
    * We ask it to add something to $2$ but not telling it what to add

In [ ]:
12 = x

* A $\rm\color{orange}{literal}$ is any value, like $12$ and $26.0$
* This is a $\rm\color{orange}{SyntaxError}$ because when Python examines that assignment statement, it knows that we cannot assign a value
to a number even before it tries to execute it
* We cannot change the value of $12$ to anything else. $12$ is just $12$

### Describing Code<br>
* Programs can be quite complicated and are often thousands of lines long
* It can be helpful to write a $\rm\color{orange}{comment}$ describing parts of the code
    * When you or someone else reads it they do not have to spend much time figuring out why the code is there
* In Python, any time the $\#$ character is encountered, Python will ignore the rest of the line
* The $\#$ symbol does not have to be the first character on the line; it can appear at the end of a statement

In [ ]:
# Python ignores this sentence because of the # symbol
print((212 - 32) * 5 / 9) # Convert 212 degrees Fahrenheit to Celsius

### Making Code Readable<br>
* Much like there are spaces in English sentences to make the words easier to read, we use spaces in Python code to make it easier to read

In [ ]:
v = 4 + -2.5 / 3.6
v=4+-2.5/3.6

* Since programs can get quite complicated, it is important that you choose names for your variables that will help you remember what they are for 
    * *id1*, *X2*, and *blah* will not remind you of anything when you come back to look at your program next week
    * Use names like *celsius*, *average*, and *final_result* instead
* It is also important to use consistent names for variables
    * If you call something *maximum* in one place, do not call it *max_val* in another
    * If you use the name *max_val*, don’t also use the name *maxVal*, and so on
---
* These rules are so important that many programming teams require members to follow a style guide for whatever language they are using
    * [PEP 8 - Style Guide for Python Code](https://peps.python.org/pep-0008/)
    * [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html)